content/learning-paths/mobile-graphics-and-gaming/android_halide/fusion.md

```diff
@@ -0,0 +1,349 @@
+blur(x, y) = Halide::cast<uint8_t>(Halide::sum(val) / 16);
+
+// Thresholding definition
+Halide::Func thresholded("thresholded");
+thresholded(x, y) = Halide::select(blur(x, y) > 128, 255, 0);
+
+// ---- Explicit use of loop fusion (fuse) ----
+Halide::Var xy("xy");  // New fused loop variable.
+
+// Fuse x and y dimensions into one linearized dimension (xy).
+thresholded.fuse(x, y, xy);
+
+// Parallelize the fused loop for improved performance.
+thresholded.parallel(xy);
+
+// Compute blur within thresholded, ensuring operation fusion.
+blur.compute_at(thresholded, xy);
+```
+
+This explicit loop-fusion example demonstrates additional optimizations achievable with Halide, clearly differentiating between operation fusion (compute_at) and loop fusion (fuse).
+
+## When to use operation fusion
+Operation fusion is especially beneficial for pipelines involving multiple sequential element-wise operations. These operations perform independent transformations on individual pixels without requiring neighboring data. Element-wise operations benefit greatly from fusion since they avoid the overhead associated with repeatedly storing and loading intermediate results, significantly reducing memory bandwidth usage.
+
+Ideal use-cases for fusion are 
+* Pixel intensity normalization (scaling and shifting)
+* Color-space transformations (e.g., RGB to grayscale conversion)
+* Simple arithmetic or logical operations applied pixel-by-pixel
+
+Here’s an illustrative example:
+```cpp
+Halide::Var x("x"), y("y");
+Halide::Func scale("scale"), offset("offset"), clamp_result("clamp_result");
+
+// Element-wise transformations
+scale(x, y) = input(x, y) * 1.5f;
+offset(x, y) = scale(x, y) + 10.0f;
+clamp_result(x, y) = Halide::clamp(offset(x, y), 0.0f, 255.0f);
+
+// Apply fusion for maximum efficiency
+scale.compute_at(clamp_result, x);
+offset.compute_at(clamp_result, x);
+```
+In this scenario, fusion is highly effective, eliminating unnecessary intermediate storage and significantly improving performance.
+
+However, fusion can introduce redundancy and inefficiencies when dealing with spatial operations such as blurs or convolutions, especially if:
+* The intermediate results are used multiple times.
+* The spatial filter has a large kernel (e.g., large Gaussian blur).
+* There are multiple sequential layers of spatial filters (e.g., multiple convolution layers).
+
+Consider a Gaussian blur with a large kernel (e.g., 15×15):
+```cpp
+Halide::Func large_blur("large_blur"), threshold("threshold");
+
+// Large Gaussian blur
+large_blur(x, y) = Halide::sum(input(x + r.x, y + r.y) * weight(r.x, r.y));
+
+// Thresholding based on blurred result
+threshold(x, y) = Halide::select(large_blur(x, y) > 128, 255, 0);
+```
+
+Using aggressive fusion here (large_blur.compute_at(threshold, x)) can result in significant computational redundancy, as each blurred value might be recomputed multiple times for overlapping pixels.
```

Comment:
This does not make sense. The consumer of blue (thresholding) is pixel-wise filter which does not need neighbor pixels, so there should be no overlapping